File: packages/contracts/test/local-node.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";
import Ajv2020 from "ajv/dist/2020.js";

test("Local Node private control fixtures obey the shared closed schema", async () => {
  const schema = JSON.parse(await readFile(new URL("../schemas/local-node/control.schema.json", import.meta.url)));
  const cases = JSON.parse(await readFile(new URL("./fixtures/local-node.json", import.meta.url)));
  const ajv = new Ajv2020({ strict: true }).addSchema(schema);
  for (const item of cases) assert.equal(ajv.getSchema(`${schema.$id}#/$defs/${item.kind}`)(item.value), item.valid, JSON.stringify(item));
});

File: packages/contracts/test/go/local_node_test.go
package contracts_test

import (
	local "convenewire.dev/contracts/generated/go/localnode"
	"encoding/json"
	"os"
	"testing"
)

func TestLocalNodeSharedFixtures(t *testing.T) {
	data, err := os.ReadFile("../fixtures/local-node.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []struct {
		Kind  string          `json:"kind"`
		Valid bool            `json:"valid"`
		Value json.RawMessage `json:"value"`
	}
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatal(err)
	}
	for _, item := range cases {
		var value any
		err := local.Decode(item.Kind, item.Value, &value)
		if (err == nil) != item.Valid {
			t.Fatalf("%s validity differs: %s", item.Kind, item.Value)
		}
	}
}
